lib/util/xml.js

```diff
@@ -6,37 +6,286 @@
+const nothing = {}; // closed sentinel value; null/undef should be valid results.
+const traverseXml = (input, initTraversers) => ExplicitPromise.of(new Promise((resolve) => {
+  // we will statefully keep track of the traversers and results over time.
+  const traversers = initTraversers.slice();
+  const results = new Array(traversers.length).fill(nothing);
+
+  // a helper for the hparser definition below.
+  const distribute = (event, x, y) => {
+    // first distribute all our events, and either accept the new traverser
+    // or write the output value to the results.
+    for (let idx = 0; idx < traversers.length; idx += 1) {
+      if (results[idx] === nothing) {
+        const result = traversers[idx](event, x, y);
+        if (typeof result === 'function')
+          traversers[idx] = result;
+        else
+          results[idx] = result;
+      }
+    }
+
+    // now, if everything is done we trip an early parsing termination.
+    if (!results.includes(nothing)) parser.parseComplete();
+  };
+
+  // define our parser and have it distribute events; it gets fed data below.
+  const parser = new hparser.Parser({
+    onopentag: (name, attrs) => { distribute('open', name, attrs); },
+    ontext: (text) => { distribute('text', text); },
+    onclosetag: () => { distribute('close'); },
+
+    // however we call end, we want to resolve with the results we did have.
+    // we crush the nothing sentinel value down to Option.none at this point.
+    onend: () => resolve(results.map((x) => (x === nothing) ? Option.none() : Option.of(x)))
+  }, { xmlMode: true });
+
+  if (typeof input.pipe === 'function') {
+    // we have a stream if .pipe() exists:
+    const decoder = new StringDecoder('utf8');
+    input.on('data', (chunk) => parser.write(decoder.write(chunk)));
+    input.on('end', () => parser.end());
+  } else {
+    // otherwise we have a string or a buffer (or something crazy that will
+    // probably crash).
+    parser.write(input);
+    parser.end();
+  }
+}));
+
+////////////////////////////////////////
+// STACKLEVEL TRAVERSERS
+// these traversers do the work of understanding the stack, and where we are in
+// it. by doing nothing more than this at a time, we make state-handling as
+// obviously-correct as we can.
+//
+// in general, these traversers receive theirs /and all nested/ events. it is
+// their job to distribute or ignore the appropriate ones.
 //
-// Various helper functions for using fast-xml-parser, with useful default options.
-// Also contains some helpers for dealing with XML processing.
+// also in general, they take filters, which are (tagname, attrs) => bool indicating
+// whether the node is a match or not.
+
+// helper for tracking real/imaginary depth traversal; if we are down a matching
+// path, we want to know how many levels deep we are. if we are partway down a
+// matching path then fail a match, we still need to know how deep we are so that
+// we understand when to go back to trying matches.
```

Comment:
This comment is very good. Maybe just going one step further? "if we are down a matching path, we want to know how many levels deep we are **so we know when to stop**"

"if we are partway down a matching path then fail a match, we **need to know how far down a non-matching path we have gone so we can identify when we see the closing event for the last matching tag we saw**"

Just some ideas.